### PR TITLE
Added ENS202EXT support

### PR DESCRIPTION
**Pull request by ntzrmtthihu777**
Works for the most part. Can't get the reset button to activate netconsole or
httpd or whatnot yet, hopefully you can point me in the right direction?

Signed-off-by: Marty Plummer <ntzrmtthihu777@gmail.com>

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -259,6 +259,7 @@ dragino_v2_ms14:
 
 p2w_cpe505n \
 p2w_r602n \
+senao-ens202ext \
```
Comment (pepe2k) [inline Makefile]: Why did you use "senao" instead of "engenius" here?

**Comment by pepe2k** [comment]
Hi @ntzrmtthihu777,

> Can't get the reset button to activate netconsole or httpd or whatnot yet, hopefully you can point me in the right direction?

Does the network work at all (ping, web server started manually with httpd command)?

**Comment by ntzrmtthihu777** [comment]
I used the name senao because that's the root company, and they have in the past had other devices marketed under differing names. Plus, the firmware from EnGenius (sysupgrade.bin format) use this naming format.

Yes, httpd/netconsole/etc all actually work when manually started. I just don't have the 'hold reset for x seconds to activate y' working.

**Comment by pepe2k** [comment]
Hi @ntzrmtthihu777,

> I used the name senao because that's the root company, and they have in the past had other devices marketed under differing names. Plus, the firmware from EnGenius (sysupgrade.bin format) use this naming format.

I prefer to use manufacture name here (and this is my general convention in this project).
BTW. try to google "Senao ENS202EXT", then "EnGenius ENS202EXT" and compare the results.

> Yes, httpd/netconsole/etc all actually work when manually started. I just don't have the 'hold reset for x seconds to activate y' working.

Are you sure that the reset button is connected to GPIO1?

You can easily test it with small script in U-Boot CLI `if button; then echo pressed; else echo not pressed; fi` (write this in CLI, hit enter, then press button and hit enter again). It should look like:

```
u-boot> if button; then echo pressed; else echo not pressed; fi
not pressed
u-boot>
pressed
u-boot>
```

Cheers,
Piotr

**Comment by pepe2k** [comment]
Ah, I forgot, show me output from `printenv`, maybe your device has already a proper U-Boot environment (and thus recovery script isn't included there).

**Comment by ntzrmtthihu777** [comment]
Ah, that's a fairly good point. I'll go ahead and put the name as EnGenius.
The test works.
And yes, the u-boot env already exists on this board, it is:
```bash
bootargs=console=ttyS0,115200 root=31:04 rootfstype=squashfs init=/etc/preinit mtdparts=ath-nor0:256k(u-boot),64k(u-boot-env),320k(custom),1536k(kernel),12096k(rootfs),2048k(failsafe),64k(ART) board=ENS202EXT
bootcmd=bootm 0x9f0a0000
baudrate=115200
factory_boot=tftp 0x3000000 vmlinux-art-ramdisk && bootm 0x3000000
failsafe_boot=bootm 0x9fdf0000
vmlinux_start_addr=0xbf0a0000
vmlinux_size=0x00150000
vmlinux_checksum=4c99d63ca5bfd89cb95f3097435637a0
rootfs_start_addr=0xbf220000
rootfs_size=0x001d0000
rootfs_checksum=17c3986dc0ee782f1ad638040dde22a6
ethaddr=88:dc:96:52:24:83
eth1addr=88:dc:96:52:24:83
serverip=192.168.1.3
ethact=eth0
ipaddr=192.168.1.10
gatewayip=192.168.1.1
stdin=serial
stdout=serial
stderr=serial
```

**Comment by pepe2k** [comment]
> And yes, the u-boot env already exists on this board, it is:
> [...]

So this is the reason, env on your board is missing few values, including the `recovery` which contains the script (example from MR3020):

`recovery=if button; then sleep 600;setenv cnt 0;setenv stop_boot;echo Keep button pressed for at least:;echo - 3s for web based recovery;echo - 5s for U-Boot console;echo - 7s for network console;echo;while button && itest $cnt < 0xA; do ledon;sleep 300;echo . '\\c';sleep 300;ledoff;sleep 600;setexpr cnt $cnt + 1;done;echo 0x$cnt seconds;echo;if itest $cnt >= 0xA; then echo \#\# Error: 10s limit reached!;echo Continuing normal boot...;echo;elif itest $cnt >= 7; then echo Starting network console...;setenv stop_boot 1;echo;startnc;elif itest $cnt >= 5; then echo Starting U-Boot console...;setenv stop_boot 1;echo;elif itest $cnt >= 3; then echo HTTP server is starting for firmware update...;setenv stop_boot 1;echo;httpd;elif itest $cnt < 3; then echo \#\# Error: button was not pressed long enough!;echo Continuing normal boot...;echo;fi;setenv cnt;fi`

I don't know how to solve this now.

One of the reasons for having recovery script inside environment value was to make it easily configurable and also make possible to remove it (in old code version it was hard coded). So, a simple solution with overwriting `recovery` value if it's empty isn't acceptable. Maybe a separated command which would fill all missing recovery related env values on demand could work here. I will think about it.

Cheers,
Piotr

File: Makefile [hunk 2]
```diff
@@ -259,6 +259,7 @@ dragino_v2_ms14:
 
 p2w_cpe505n \
 p2w_r602n \
+engenius-ens202ext \
```
Comment (pepe2k) [inline Makefile]: Use `engenius_ens202ext` here, please.

File: u-boot/Makefile [hunk 3]
```diff
@@ -470,6 +470,13 @@ p2w_r602n: qca953x_common
 	@$(call define_add,CFG_ATH_GMAC_NMACS,2)
 	@$(MKCONFIG) -a ap143 mips mips ap143 ar7240 ar7240
 
+engenius-ens202ext: ar934x_common
+	@$(call config_init,Senao ENS202EXT,ens202ext,16,1,1,QCA_AR9341_SOC)
```
Comment (pepe2k) [inline u-boot/Makefile]: Senao ENS202EXT -> EnGenius ENS202EXT

File: u-boot/Makefile [hunk 4]
```diff
@@ -470,6 +470,13 @@ p2w_r602n: qca953x_common
 	@$(call define_add,CFG_ATH_GMAC_NMACS,2)
 	@$(MKCONFIG) -a ap143 mips mips ap143 ar7240 ar7240
 
+engenius-ens202ext: ar934x_common
```
Comment (pepe2k) [inline u-boot/Makefile]: engenius_ens202ext here, please.

**Comment by ntzrmtthihu777** [comment]
@pepe2k could those not be stored as plain u-boot env stuffs in the u-boot-env mtd partition?
If it doesn't have access to that early on I would not be opposed to it, I just need to know if it needs
to be within the first u-boot mtd partition or not.

**Comment by pepe2k** [comment]
@ntzrmtthihu777,

> @pepe2k could those not be stored as plain u-boot env stuffs in the u-boot-env mtd partition?

Values stored there are used during low level pll/clock initialization (written in [asm](https://github.com/pepe2k/u-boot_mod/blob/master/u-boot/cpu/mips/ar7240/qca95xx_pll_init.S) at this moment), so the answer is no.

> If it doesn't have access to that early on I would not be opposed to it, I just need to know if it needs
> to be within the first u-boot mtd partition or not.

I usually use part of the environment sector (which also makes the env smaller). To be able to write there, you need to erase whole sector in advance (NOR flash), which could be dangerous if you use e.g. art data partition for that.

As I'm thinking about it now... maybe it would make sense to store PLL configuration somewhere else (just after the U-Boot or at the end of ART data) and use smaller sector erase (4k) instead of whole block (64k). That wouldn't break U-Boot environment compatibility/support (size would match) in firmware (LEDE/OpenWrt).

Cheers,
Piotr

**Comment by pepe2k** [comment]
Merged locally (with some fixes/changes), thanks.

**Pull request closed without merging**